Fix interpolation error using Discrete Sliders

lucaantonelli opened the pull request:
Fixes #39510
Fixes #73787

This PR fix an interpolation error with some values specified in the related issue. I modified the lerp function in Slider, RangeSlider and CupertinoSlider. Also add the ui import for lerpDouble in material Slider to be more consistent with CupertinoSlider that uses the same approach. Add tests for all the 3 sliders.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [x] All existing and new tests are passing.


<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#overview
[Tree Hygiene]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md
[test-exempt]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md
[Features we expect every widget to implement]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/blob/main/docs/contributing/Chat.md
[Data Driven Fixes]: https://github.com/flutter/flutter/blob/main/docs/contributing/Data-driven-Fixes.md

lucaantonelli committed:
Fix interpolation error in Sliders values with divisions set

dkwingsmt commented on packages/flutter/lib/src/material/range_slider.dart:
```diff
@@ -573,7 +573,13 @@ class _RangeSliderState extends State<RangeSlider> with TickerProviderStateMixin
   // be between 0.0 and 1.0.
-  double _lerp(double value) => ui.lerpDouble(widget.min, widget.max, value)!;
+  double _lerp(double value) {
+    double lerpValue = ui.lerpDouble(widget.min, widget.max, value)!;
+    if (widget.divisions != null) {
+      lerpValue = (lerpValue * widget.divisions!).round() / widget.divisions!;
+    }
```
Rather than computing `lerpValue` twice, maybe put it in a `if...else` statement with two returns?

dkwingsmt commented on packages/flutter/lib/src/material/slider.dart:
```diff
@@ -765,7 +766,12 @@ class _SliderState extends State<Slider> with TickerProviderStateMixin {
     assert(value <= 1.0);
-    return value * (widget.max - widget.min) + widget.min;
+
+    double lerpValue = lerpDouble(widget.min, widget.max, value)!;
+    if (widget.divisions != null) {
+      lerpValue = (lerpValue * widget.divisions!).round() / widget.divisions!;
+    }
+    return lerpValue;
```
Same, don't compute it twice.

dkwingsmt reviewed (commented):
I'm not convinced of this approach. We can not simply apply `round` to general cases, because it is not guaranteed that `min` and `max` are integers and their difference can be divided by `divisions`. 

Maybe we can add a boolean called something like `integralDivisions` to let developers choose whether to round the division points.

dkwingsmt commented:
Hi! @lucaantonelli  Are you able to address the comments above so we can work together to get this landed? This is a great addition and I'm looking forward to it!

lucaantonelli commented:
> Hi! @lucaantonelli  Are you able to address the comments above so we can work together to get this landed? This is a great addition and I'm looking forward to it!

Hi! I'm working on it, i will commit in this days :)

lucaantonelli commented:
Hi @dkwingsmt , I implemented the new integralDivisions param with a doc, and fixed the return statements in the two material sliders. Let me now if something else could be improved!

dkwingsmt reviewed (commented):
Left some comments. The same applies to Material.

lucaantonelli commented on packages/flutter/lib/src/cupertino/slider.dart:
```diff
@@ -235,7 +243,10 @@ class CupertinoSlider extends StatefulWidget {
-    final double lerpValue = lerpDouble(widget.min, widget.max, value)!;
+    double lerpValue = lerpDouble(widget.min, widget.max, value)!;
+    if (widget.integralDivisions && widget.divisions != null) {
+      lerpValue = (lerpValue * widget.divisions!).round() / widget.divisions!;
```
I choose your version, mine was covering more cases also for non integer values but wasn't perfect in some edge cases. I just wonder if an assert could be useful, since if someone enables integralDivisions but put a non integer value could lead to assert error only at runtime, with something like that at widget level we could prevent this:

```
assert(
  !integralDivisions || divisions == null || (max - min) % divisions == 0.0,
  'When integralDivisions is true, (max - min) must be exactly divisible by divisions. ',
)
```

I don't know if putting an operations in assert is not a good standard, or at all we don't want to catch this error and let the user face it by himself. 

Example to explain me better: if i put 5 divisions and than min = 0 and max = 20.5, the last value of the slider will be 21 if integralDivisions is true so throwing assert error when moving slider to the end, let me know what do you think :)

dkwingsmt commented:
Continuing the discussion in https://github.com/flutter/flutter/pull/171067#discussion_r2297715362 in case the thread is buried: 

That's a really good point!

I also thought about the divisibility test, and I think the core question is: This assertion is probably viable for like 90% of the usages of `integralDivisions`, but should we sacrifice the last 10%? Adding this assertion doesn't make our implementation easier, and doesn't change the result of cases where the range _is_ divisible. Therefore I suggest we don't add this assertion for now.

The last paragraph, though, speaks a real problem: the last value of the slider must not be larger than `max` in any cases. So if `min = 0.1, max = 10.8, integralDivisions = true`, then:
* The first value should be 1.0, and the last value should be 10.0. 
* The dividing points in between should be equally divided within [1.0, 10.0] (not [0.1, 10.8]).
* The full range of the slider should still be [0.1, 10.8], i.e. the thumb will never reach the tip of either end.

(Although, it seems to me that satisfying the design above requires non-trivial refactors due to how `_discretizedCurrentDragValue` is designed.)

dkwingsmt commented:
Hi @lucaantonelli Have you had opportunity to look at my comment above? (from triage)

victorsanni committed:
Merge branch 'master' into fix-slider-discrete-value-precision
